host/workflowsidinternalratelimit_test.go

```diff
@@ -0,0 +1,230 @@
+		if activityCounter < activityCount {
+			activityCounter++
+			buf := new(bytes.Buffer)
+			s.Nil(binary.Write(buf, binary.LittleEndian, activityCounter))
+
+			return []byte(strconv.Itoa(int(activityCounter))), []*types.Decision{{
+				DecisionType: types.DecisionTypeScheduleActivityTask.Ptr(),
+				ScheduleActivityTaskDecisionAttributes: &types.ScheduleActivityTaskDecisionAttributes{
+					ActivityID:                    strconv.Itoa(int(activityCounter)),
+					ActivityType:                  &types.ActivityType{Name: activityName},
+					TaskList:                      &types.TaskList{Name: testTaskListName},
+					Input:                         buf.Bytes(),
+					ScheduleToCloseTimeoutSeconds: common.Int32Ptr(1),
+					ScheduleToStartTimeoutSeconds: common.Int32Ptr(1),
+					StartToCloseTimeoutSeconds:    common.Int32Ptr(1),
+					HeartbeatTimeoutSeconds:       common.Int32Ptr(1),
+				},
+			}}, nil
+		}
+
+		s.Logger.Info("Completing Workflow.")
+
+		workflowComplete = true
+		return []byte(strconv.Itoa(int(activityCounter))), []*types.Decision{{
+			DecisionType: types.DecisionTypeCompleteWorkflowExecution.Ptr(),
+			CompleteWorkflowExecutionDecisionAttributes: &types.CompleteWorkflowExecutionDecisionAttributes{
+				Result: []byte("Done."),
+			},
+		}}, nil
+	}
+
+	activityExecutedCount := int32(0)
+	atHandler := func(execution *types.WorkflowExecution, activityType *types.ActivityType,
+		activityID string, input []byte, taskToken []byte) ([]byte, bool, error) {
+		s.Equal(testWorkflowID, execution.WorkflowID)
+		s.Equal(activityName, activityType.Name)
+
+		activityExecutedCount++
+		return []byte("Activity Result."), false, nil
+	}
+
+	poller := &TaskPoller{
+		Engine:          s.engine,
+		Domain:          s.domainName,
+		TaskList:        &types.TaskList{Name: testTaskListName},
+		Identity:        testIdentity,
+		DecisionHandler: dtHandler,
+		ActivityHandler: atHandler,
+		Logger:          s.Logger,
+		T:               s.T(),
+	}
+
+	for i := int32(0); i < activityCount; i++ {
+		_, err = poller.PollAndProcessDecisionTask(false, false)
+		s.Nil(err)
+
+		err = poller.PollAndProcessActivityTask(false)
+		s.Nil(err)
+	}
+
+	s.Logger.Info("Waiting for workflow to complete", tag.WorkflowRunID(we.RunID))
+
+	s.False(workflowComplete)
+	_, err = poller.PollAndProcessDecisionTask(false, false)
+	s.Nil(err)
+	s.True(workflowComplete)
+
+	historyResponse, err := s.engine.GetWorkflowExecutionHistory(ctx, &types.GetWorkflowExecutionHistoryRequest{
+		Domain: s.domainName,
+		Execution: &types.WorkflowExecution{
+			WorkflowID: testWorkflowID,
+		},
+	})
+	s.Nil(err)
+	history := historyResponse.History
+	firstEvent := history.Events[0]
+	lastEvent := history.Events[len(history.Events)-1]
+	// First 7 event ids --> 0 (Workflow start), 1-3 ( Decision scheduled,started,completed) , 4-6 (Activity scheduled,started,completed) post which the tasks will be rate limited since RPS is 2
```

Comment:
it's not super clear to me how this test validates workflow id rate limits. how do you ensure these specific events are spread with 2 rps?

Reply:
These events correspond to two tasks - one decision task and one activity task. The ratelimiter is at the task processing level and after processing 2 tasks, the rate limiter starts to return an error

Reply:
I guess there's no externally visible error being returned, just a delay?

Reply:
yes there is no externally visible error since the rate limit error is being retried and delays the task processing